### PR TITLE
Integration of the skrl RL library

### PR DESCRIPTION
**Pull request by Toni-SM**
# Description

Integration of the **skrl** reinforcement learning library 

* Adds a wrapper function to wrap an IsaacEnv for skrl compatibility
* Adds a trainer class that also logs episode information
* Adds train.py and play.py in workflows to run skrl
* Adds configuration files for currently included environments with parameters matching that of rl_games as close as possible

Dependencies: 
- skrl >= 0.10.0: https://github.com/Toni-SM/skrl

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see [here](https://pre-commit.com/#install) instructions to set it up)
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file

### CODE REVIEW

**Comment by Mayankm96** [comment]
Thanks a lot for the pull request! I'll take a look at it tomorrow.

Can you please also update the changelog and the version in the `extension.toml` file of `omni.isaac.orbit_envs`?

> https://isaac-orbit.github.io/orbit/source/refs/contributing.html#maintaining-a-changelog

**Comment by Toni-SM** [comment]
Done, CHANGELOG and extension.toml files updated!!!

**Comment by Mayankm96** [comment]
Hi!

I finally got time today to test this out. I am yet to try training on the classic environments but would be nice to do some evaluations while having the same configurations for PPO across the frameworks.

Right now, mostly tried training for the reach environment. I have the following comments.

## Issues

1. Currently, while training the reach environment, I am seeing a divergence in the loss curve of the value function. Is it possible the parameters are not tuned? I suppose there might be tiny differences in algorithmic implementations but don't think the value function should be diverging. One possible explanation could be that the learning iterations are too large and that's causing some kind of collapse.

     value_skrl

2. The learned policy for reach seems to be not that stable (I tried the `best_agent.pt`) . It is quite shaky which I don't see when I train with rsl_rl. Maybe again the parameters are not similar or the same and that is the potential problem.

    https://user-images.githubusercontent.com/12863862/214456059-a4a6774f-5114-4bae-8ec5-fbda6e125307.mp4

3. Currently, the play.py is using the runner class to run the agent-env interaction. 
    * When I hit the `STOP` icon on the GUI, it just freezes and doesn't exit the script quickly. Is there a possible issue in the trainer that doesn't let it exit? 
    * When I hit the `X` icon on the GUI, I observe a similar behavior.

## Nice to haves

* Is it possible to also log the episode information in the runner? Many times in robotics, we care about how each reward term is evolving (for weight tuning). The cumulative reward is often hard to decipher when having many reward terms
* Can the `play.py` just load the last or best checkpoint by default? We do something similar for rsl-rl ([code snippet](https://github.com/NVIDIA-Omniverse/Orbit/blob/main/source/standalone/workflows/rsl_rl/play.py#L63-L77)).

## Additional comments

* We will also need to add the license of `skrl` inside the `docs/licenses/dependencies directory for book-keeping purposes.

**Comment by Toni-SM** [comment]
Hi @Mayankm96 

<hr>
Comments:

1. Loss curve divergence of the value function should not be a problem. rl_games also diverge... I checked the rsl_rl PPO implementation and there is a small difference in the way KL and learning rate scheduler are computed, also some training parameters are different

    Screenshot from 2023-01-25 17-49-14

2. I mapped (as far as possible) the same parameters as .yaml files for rl_games. The stability of the policy comes mostly from the number of time steps used compared to rsl_rl. 
    
    - skrl: 8000, 
    - rl_games: 16 * 500 = 8000
    - rsl_rl:  64 * 250 = 16000
 
    rls_rl works just as "badly" as the others at 8000 timesteps.
    Training for 16000 timesteps generates a better curves

   Screenshot from 2023-01-25 18-17-24

3. **Solved**

<hr>
Nice to haves

1. Log the episode information: 
    **Done**

2. Load the last or best checkpoint by default: 
    **Done**

3. An updated example of how to use a learned policy outside the runner loop:
    It sounds great. I am going to create an example for futures pull requests

<hr>

Additional comments

1. skrl license in docs: 
    **Done**
<br>

File: source/standalone/workflows/skrl/config.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    def reward_shaper_function(scale):
+        def reward_shaper(rewards, timestep, timesteps):
+            return rewards * scale
+
+        return reward_shaper
+
+    def update_dict(d):
+        from skrl.resources.preprocessors.torch import RunningStandardScaler  # noqa: F401
+        from skrl.resources.schedulers.torch import KLAdaptiveRL  # noqa: F401
+        from skrl.utils.model_instantiators import Shape  # noqa: F401
```
Comment (Mayankm96) [inline source/standalone/workflows/skrl/config.py]: Is it alright to move these imports to the top of the file instead? Mainly to follow the PEP convention as much as possible.

File: source/standalone/workflows/skrl/train.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+                    obs, infos = env.reset()
+                else:
+                    obs.copy_(next_obs)
+            # check if simulator is stopped
+            if env.sim.is_stopped():
+                break
+
+        # close the simulator
+        env.close()
+        simulation_app.close()
```
Comment (Mayankm96) [inline source/standalone/workflows/skrl/train.py]: Do we need this more custom version of the trainer? For training, it is fine to use a trainer, because we assume the user would handle their custom trainers from the library side and not over here. 

I feel right now it adds some bulk and makes the script longer than it needs to be. Maybe I am missing something here.

File: source/standalone/workflows/skrl/play.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+        # reset environment
+        obs, infos = env.reset()
+        # simulate environment
+        while simulation_app.is_running():
+            timestep += 1
+            # agent stepping
+            actions = agent.act(obs, timestep=timestep, timesteps=None)[0]
+            # env stepping
+            next_obs, rewards, terminated, truncated, infos = env.step(actions)
```
Comment (Mayankm96) [inline source/standalone/workflows/skrl/play.py]: After every env step call, it is important to check if the simulator is stopped or not and exit right away. The main reason is that, in Isaac Sim, if the user presses stop, then all the physics handles are disengaged and we can no longer read any new data from the physics buffers. Thus, inside the `IsaacEnv` class, we check for this and return the last values that were stored in the buffers. This logic is defined [here](https://github.com/NVIDIA-Omniverse/Orbit/blob/main/source/extensions/omni.isaac.orbit_envs/omni/isaac/orbit_envs/isaac_env.py#L246-L275).

Thus, as a practice, I would suggest something similar to [here](https://github.com/NVIDIA-Omniverse/Orbit/blob/main/source/standalone/workflows/rl_games/play.py#L128-L132):

```python
# env stepping
obs, _, dones, _ = env.step(actions)
# check if simulator is stopped
if env.unwrapped.sim.is_stopped():
    break
```

File: source/standalone/workflows/skrl/play.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+                timestep=timestep,
+                timesteps=None,
+            )
+            # log custom environment data
+            if "episode" in infos:
+                for k, v in infos["episode"].items():
+                    if isinstance(v, torch.Tensor) and v.numel() == 1:
+                        agent.track_data(f"Info / {k}", v.item())
+            # write data to TensorBoard / Weights & Biases
+            super(type(agent), agent).post_interaction(timestep=timestep, timesteps=None)
```
Comment (Mayankm96) [inline source/standalone/workflows/skrl/play.py]: Do we need logging when playing?

File: source/standalone/workflows/skrl/play.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+            # reset environments
+            if terminated.any() or truncated.any():
+                obs, infos = env.reset()
+            else:
+                obs.copy_(next_obs)
```
Comment (Mayankm96) [inline source/standalone/workflows/skrl/play.py]: Is this calling reset over all the environments? Normally that is handled within the step call (similar to in isaacgym).

**Review by Mayankm96 — CHANGES_REQUESTED** [review]
Thanks a lot for providing more information. I just added some minor feedback on the code to understand things a bit better. Once that is cleared up, we can merge this awesome PR :)

File: source/standalone/workflows/skrl/config.py (hunk 1 above)
Comment (Toni-SM) [reply source/standalone/workflows/skrl/config.py]: Yes, it is possible... Their import is necessary to convert some configurations described in .yaml to classes expected by the library via the Python `eval(...)` method

File: source/standalone/workflows/skrl/train.py (hunk 2 above)
Comment (Toni-SM) [reply source/standalone/workflows/skrl/train.py]: Well, running all those steps to train/evaluate the agent is the only way to track custom data (Episode separated rewards)...
Because the `''episode"` key in the `info` dictionary is not an standard, library code does not track It by default

File: source/standalone/workflows/skrl/play.py (hunk 3 above)
Comment (Toni-SM) [reply source/standalone/workflows/skrl/play.py]: This is called at the end of each iteration loop:

```python
            # check if simulator is stopped
            if env.sim.is_stopped():
                break
```

All methods that are called after `env.step` and before this, belong to agent and don't interact with the environment

File: source/standalone/workflows/skrl/play.py (hunk 4 above)
Comment (Toni-SM) [reply source/standalone/workflows/skrl/play.py]: No... I just find it useful. But it can be removed without any problem

File: source/standalone/workflows/skrl/play.py (hunk 5 above)
Comment (Toni-SM) [reply source/standalone/workflows/skrl/play.py]: For a parallel environment, it does not reset the environments (except for its first calls). 
The method is defined as just for compatibility with the Gym interface... skrl manage several interfaces

File: source/standalone/workflows/skrl/train.py (hunk 2 above)
Comment (Toni-SM) [reply source/standalone/workflows/skrl/train.py]:
```python
            # log custom environment data
            if "episode" in infos:
                for k, v in infos["episode"].items():
                    if isinstance(v, torch.Tensor) and v.numel() == 1:
                        agent.track_data(f"Info / {k}", v.item())
```

File: source/standalone/workflows/skrl/play.py (hunk 3 above)
Comment (Toni-SM) [reply source/standalone/workflows/skrl/play.py]: But I can move it just after the environment stepping

File: source/standalone/workflows/skrl/play.py (hunk 5 above)
Comment (Mayankm96) [reply source/standalone/workflows/skrl/play.py]: Aha I see. It is mostly from the trainer class itself. Makes sense now :)

File: source/standalone/workflows/skrl/play.py (hunk 3 above)
Comment (Mayankm96) [reply source/standalone/workflows/skrl/play.py]: Yeah didn't realize this was just how the trainer class works. Makes sense now for consistent behavior.

**Comment by Toni-SM** [comment]
Done...

Summary:

- Create a customized trainer in the file where the `skrl` wrapper is defined for: 
     - log episode information during training
     - simplify `train.py` and `play.py` 
- Remove logging in `play.py`
- Move the import statements to the begging of the `config.py` file

**Review by Mayankm96 — APPROVED** [review]
The PR looks good to me now. Thanks a lot for working on skrl integration. It is great to have more libraries supported in the workflows.

Would love to try out other agents as well in skrl at some point and provide examples :)

**Comment by Toni-SM** [comment]
Great!
Let's work on integrating new agents in future pull requests!

By the way... The online documentation does not reflect any changes 🤔

**Comment by Mayankm96** [comment]
It should be up now! :)

By the way, I made some minor refactoring in this commit f7e4183d2c4bb6c56bd50166c5fc40a887f3a5ec :
* renaming of `SkrlLogTrainer` to `SkrlSequentialLogTrainer`
* readding of the eval mode to the class for completeness
* some docstring fixes
* setting the agent to "eval" mode in the play.py script

**Comment by Toni-SM** [comment]
Nice... It looks much better now

**Comment by Ozzey** [comment]
Thank you for the amazing work. Not sure if it is the correct thread to ask the question but when I resume training from a saved checkpoint in SKRL, the episode counter in training logs resets to 0 (e.g., showing 'Episode 0') instead of continuing from the last saved episode number like it does in rl_games. Does this reset mean:
a) Training is truly restarting from scratch, discarding learned progress?
b) Or is this just a logging display issue while actual training continuity is preserved?
